# HG changeset patch
# User Kai T. Ohlhus <k.ohlhus@gmail.com>
# Date 1521666147 -3600
#      Wed Mar 21 22:02:27 2018 +0100
# Branch stable
# Node ID 1fbd8afcf625866d654a3ee2bd1ad03cd4d255ba
# Parent  27d68d7a482ccb2b7c863ae1c6696898e36a7191
Ensure that local functions do not have a static workspace (bug #53405).

* libinterp/corefcn/symscope.cc: Check for nested function, before making the
workspace static.

* test/local-functions/local_function_eval.m: New test function.
* test/local-functions/local_functions.tst: New tests for local functions.
* test/local-functions/module.mk: Add files to test suite.

* test/module.mk: Add files to test suite.

diff --git a/libinterp/corefcn/symscope.cc b/libinterp/corefcn/symscope.cc
--- a/libinterp/corefcn/symscope.cc
+++ b/libinterp/corefcn/symscope.cc
@@ -185,17 +185,18 @@ namespace octave
                 && m_is_nested && t_parent->look_nonlocal (nm_sr.first, ours))
               {
                 if (ours.is_global () || ours.is_persistent ())
                   error ("global and persistent may only be used in the topmost level in which a nested variable is used");
               }
           }
 
         // The scopes of nested functions are static.
-        m_is_static = true;
+        if (m_is_nested)
+          m_is_static = true;
       }
     else if (m_children.size ())
       {
         // Parents of nested functions have static scopes.
         m_is_static = true;
       }
 
     for (auto& scope_obj : m_children)
diff --git a/test/local-functions/local_function_eval.m b/test/local-functions/local_function_eval.m
new file mode 100644
--- /dev/null
+++ b/test/local-functions/local_function_eval.m
@@ -0,0 +1,9 @@
+function status = local_function_eval ()
+  eval ("y = 5;");
+  status = local_fcn ("y = 6;");
+endfunction
+
+function status = local_fcn (expr)
+  eval (expr);
+  status = exist ("y", "var");
+endfunction
diff --git a/test/local-functions/local_functions.tst b/test/local-functions/local_functions.tst
new file mode 100644
--- /dev/null
+++ b/test/local-functions/local_functions.tst
@@ -0,0 +1,25 @@
+## Copyright (C) 2018 Kai T. Ohlhus
+##
+## This file is part of Octave.
+##
+## Octave is free software: you can redistribute it and/or modify it
+## under the terms of the GNU General Public License as published by
+## the Free Software Foundation, either version 3 of the License, or
+## (at your option) any later version.
+##
+## Octave is distributed in the hope that it will be useful, but
+## WITHOUT ANY WARRANTY; without even the implied warranty of
+## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+## GNU General Public License for more details.
+##
+## You should have received a copy of the GNU General Public License
+## along with Octave; see the file COPYING.  If not, see
+## <https://www.gnu.org/licenses/>.
+
+################################################################################
+## This file actually executes the tests on functions with local functions.
+##
+## It relies on the function files defined in the local-functions/ directory.
+################################################################################
+
+%!assert (local_function_eval (), 1)
diff --git a/test/local-functions/module.mk b/test/local-functions/module.mk
new file mode 100644
--- /dev/null
+++ b/test/local-functions/module.mk
@@ -0,0 +1,5 @@
+nest_TEST_FILES = \
+  %reldir%/subfunctions.tst \
+  %reldir%/subfunction_eval.m
+
+TEST_FILES += $(nest_TEST_FILES)
diff --git a/test/module.mk b/test/module.mk
--- a/test/module.mk
+++ b/test/module.mk
@@ -66,16 +66,17 @@ include %reldir%/bug-51599/module.mk
 include %reldir%/bug-52075/module.mk
 include %reldir%/bug-52722/module.mk
 include %reldir%/class-concat/module.mk
 include %reldir%/classdef/module.mk
 include %reldir%/classdef-multiple-inheritance/module.mk
 include %reldir%/classes/module.mk
 include %reldir%/ctor-vs-method/module.mk
 include %reldir%/fcn-handle-derived-resolution/module.mk
+include %reldir%/local-functions/module.mk
 include %reldir%/nest/module.mk
 include %reldir%/publish/module.mk
 
 define run-octave-tests
   ( cd %reldir% && $(SHELL) ../run-octave $(RUN_OCTAVE_OPTIONS) $(1) --norc --silent --no-history $(abs_top_srcdir)/%reldir%/fntests.m $(abs_top_srcdir)/%reldir% ); \
   if $(AM_V_P); then \
     echo ""; \
     if [ -f %reldir%/fntests.log ]; then \
